### PR TITLE
feat(datasource): `sourceUrl` & `releaseTimestamp` support

### LINKED ISSUE #28947: Document sourceUrl and releaseTimestamp support in datasources
### Describe the proposed change(s).

Extend https://github.com/renovatebot/renovate/blob/5de7eee2b3f7011397ea273d1289933c343da9cb/lib/modules/datasource/types.ts#L111-L114

To be like:

```
  /**
   * Whether custom registryUrls are allowed.
   */
  customRegistrySupport: boolean;

  /**
   * Whether release timestamps can be returned.
   */
   releaseTimestampSupport: boolean;
   releaseTimeStampNote?: string;

  /**
   * Whether sourceURLs can be returned.
   */
   sourceUrlSupport: 'package' | 'release' | 'none';
   sourceUrlNote?: string;
```

`releaseTimestampSupport` should be `true` if the datasources looks for release.releaseTimestamp

`sourceUrlSupport` should be `release` if the datasource supports returning them per-release (e.g. npm), `package` if they can be returned but only per-package (e.g. cdnjs), or `none` if no `sourceUrl` logic exists.

The optional note fields should ideally be filled into clarify how the field is determined. e.g in both npm and cdnjs, it's from the `repository` value in the results. Can we possibly use types to make these note fields _mandatory_ if their sibling field is set?

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes
- Added `sourceUrlSupport/Note` and `releaseTimestampSupport/Note` fields based on the return object of each datasource

## Context
Closes: https://github.com/renovatebot/renovate/issues/28947

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/datasource/conan/index.ts [hunk 1]
```diff
@@ -38,6 +38,10 @@ export class ConanDatasource extends Datasource {
 
   githubHttp: GithubHttp;
 
+  static readonly sourceUrlSupport = 'package';
```
Comment (viceice) [inline lib/modules/datasource/conan/index.ts:41]: why static?

Comment (RahulGautamSingh) [reply lib/modules/datasource/conan/index.ts:41]: left by mistake

File: lib/modules/datasource/datasource.ts [hunk 2]
```diff
@@ -25,6 +26,12 @@ export abstract class Datasource implements DatasourceApi {
 
   registryStrategy: RegistryStrategy | undefined = 'first';
 
+  releaseTimestampSupport = false;
+  releaseTimeStampNote: string | undefined;
```
Comment (secustor) [inline lib/modules/datasource/datasource.ts:30]:
```suggestion
  releaseTimeStampNote?: string | undefined;
```

File: lib/modules/datasource/dotnet-version/index.ts [hunk 3]
```diff
@@ -23,6 +23,14 @@ export class DotnetVersionDatasource extends Datasource {
     'https://dotnetcli.blob.core.windows.net/dotnet/release-metadata/releases-index.json',
   ];
 
+  // De-comment when supporting PR which adds this feature is merged into main
```
Comment (secustor) [inline lib/modules/datasource/dotnet-version/index.ts:26]: Please add PR number or link here for reference.

File: lib/modules/datasource/dotnet-version/index.ts [hunk 4]
```diff
@@ -23,6 +23,14 @@ export class DotnetVersionDatasource extends Datasource {
...
+  // override releaseTimestampSupport = true;
+  // override releaseTimeStampNote =
+  //   'The release timestamp is determined from the `release-date` field in the results.';
+  override readonly sourceUrlSupport = 'package';
+  override readonly sourceUrlNote =
+    'We use the url <https://github.com/dotnet/sdk> for `dotnet-sdk` and, <https://github.com/dotnet/runtime> for the `dotnet-runtime`.';
```
Comment (zharinov) [inline lib/modules/datasource/dotnet-version/index.ts:32]: Let's not enclose urls inside angle brackets, let it be close to Markdown

```suggestion
    'We use the url https://github.com/dotnet/sdk for `dotnet-sdk` and, https://github.com/dotnet/runtime for the `dotnet-runtime`.';
```

File: lib/modules/datasource/unity3d/index.ts [hunk 5]
```diff
@@ -24,6 +24,10 @@ export class Unity3dDatasource extends Datasource {
 
   override readonly registryStrategy = 'merge';
 
+  override readonly releaseTimestampSupport = true;
+  override readonly releaseTimeStampNote =
```
Comment (zharinov) [inline lib/modules/datasource/unity3d/index.ts:28]: Timestamp is the common term, it doesn't need to be written with separate words

```suggestion
  override readonly releaseTimestampNote =
```

**Review by zharinov — CHANGES_REQUESTED** [review]

File: lib/modules/datasource/datasource.ts (hunk 2 above)
Comment (rarkins) [reply lib/modules/datasource/datasource.ts:30]: Does it need the `| undefined` part? If so, what's the difference?

Comment (secustor) [reply lib/modules/datasource/datasource.ts:30]: It is effectively the same, I like simply to be more verbose when in doubt.

File: lib/modules/datasource/galaxy/index.ts [hunk 6]
```diff
@@ -19,6 +19,13 @@ export class GalaxyDatasource extends Datasource {
 
   override readonly defaultVersioning = pep440Versioning.id;
 
+  override readonly releaseTimestampSupport = true;
+  override readonly releaseTimestampNote =
+    'The release timestamp is determined from the `created` field in the results.';
+  override readonly sourceUrlSupport = 'package';
+  override readonly sourceUrlNote =
+    'The source url is determined from the `github_user` and `github_repo` fields in the results.';
```
Comment (HonkingGoose) [inline lib/modules/datasource/galaxy/index.ts:27]:
```suggestion
    'The source URL is determined from the `github_user` and `github_repo` fields in the results.';
```

File: lib/modules/datasource/git-tags/index.ts [hunk 7]
```diff
@@ -11,6 +11,9 @@ export class GitTagsDatasource extends GitDatasource {
   }
 
   override readonly customRegistrySupport = false;
+  override readonly sourceUrlSupport = 'package';
+  override readonly sourceUrlNote =
+    'The source url is determined by using the `packageName` and `registryUrl`.';
```
Comment (HonkingGoose) [inline lib/modules/datasource/git-tags/index.ts:16]:
```suggestion
    'The source URL is determined by using the `packageName` and `registryUrl`.';
```

File: lib/modules/datasource/gitea-releases/index.ts [hunk 8]
```diff
@@ -16,6 +16,13 @@ export class GiteaReleasesDatasource extends Datasource {
 
   private static readonly cacheNamespace: PackageCacheNamespace = `datasource-${GiteaReleasesDatasource.id}`;
 
+  override readonly releaseTimestampSupport = true;
+  override readonly releaseTimestampNote =
+    'The release timestamp is determined from the `published_at` field in the results.';
+  override readonly sourceUrlSupport = 'package';
+  override readonly sourceUrlNote =
+    'The source url is determined by using the `packageName` and `registryUrl`.';
```
Comment (HonkingGoose) [inline lib/modules/datasource/gitea-releases/index.ts:24]:
```suggestion
    'The source URL is determined by using the `packageName` and `registryUrl`.';
```

File: lib/modules/datasource/gitea-tags/index.ts [hunk 9]
```diff
@@ -16,6 +16,13 @@ export class GiteaTagsDatasource extends Datasource {
 
   private static readonly cacheNamespace: PackageCacheNamespace = `datasource-${GiteaTagsDatasource.id}`;
 
+  override readonly releaseTimestampSupport = true;
+  override readonly releaseTimestampNote =
+    'The release timestamp is determined from the `created` field in the results.';
+  override readonly sourceUrlSupport = 'package';
+  override readonly sourceUrlNote =
+    'The source url determined by using the `packageName` and `registryUrl`.';
```
Comment (HonkingGoose) [inline lib/modules/datasource/gitea-tags/index.ts:24]:
```suggestion
    'The source URL is determined by using the `packageName` and `registryUrl`.';
```

This seems to be the correct pattern?

File: lib/modules/datasource/github-release-attachments/index.ts [hunk 10]
```diff
@@ -38,6 +38,14 @@ export class GithubReleaseAttachmentsDatasource extends Datasource {
 
   override http: GithubHttp;
 
+  override readonly releaseTimestampSupport = true;
+  // Note: not sure
+  override readonly releaseTimestampNote =
+    'The release timestamp is determined from the `releaseTimestamp` field in the results.';
+  override readonly sourceUrlSupport = 'package';
+  override readonly sourceUrlNote =
+    'The source url is determined by using the `packageName` and `registryUrl`.';
```
Comment (HonkingGoose) [inline lib/modules/datasource/github-release-attachments/index.ts:47]:
```suggestion
    'The source URL is determined by using the `packageName` and `registryUrl`.';
```

File: lib/modules/datasource/github-releases/index.ts [hunk 11]
```diff
@@ -21,6 +21,14 @@ export class GithubReleasesDatasource extends Datasource {
 
   override http: GithubHttp;
 
+  override readonly releaseTimestampSupport = true;
+  // Note: not sure
+  override readonly releaseTimestampNote =
+    'The release timestamp is determined from the `releaseTimestamp` field from the response.';
+  override readonly sourceUrlSupport = 'package';
+  override readonly sourceUrlNote =
+    'The source url is determined by using the `packageName` and `registryUrl`.';
```
Comment (HonkingGoose) [inline lib/modules/datasource/github-releases/index.ts:30]:
```suggestion
    'The source URL is determined by using the `packageName` and `registryUrl`.';
```

File: lib/modules/datasource/github-runners/index.ts [hunk 12]
```diff
@@ -5,6 +5,10 @@ import type { GetReleasesConfig, Release, ReleaseResult } from '../types';
 export class GithubRunnersDatasource extends Datasource {
   static readonly id = 'github-runners';
 
+  override readonly sourceUrlSupport = 'package';
+  override readonly sourceUrlNote =
+    'We use the url https://github.com/actions/runner-images.';
```
Comment (HonkingGoose) [inline lib/modules/datasource/github-runners/index.ts:10]:
```suggestion
    'We use the URL: https://github.com/actions/runner-images.';
```

File: lib/modules/datasource/github-tags/index.ts [hunk 13]
```diff
@@ -20,6 +20,14 @@ export class GithubTagsDatasource extends Datasource {
 
   override readonly registryStrategy = 'hunt';
 
+  override readonly releaseTimestampSupport = true;
+  // Note: not sure
+  override readonly releaseTimestampNote =
+    'The get release timestamp is determined from the `releaseTimestamp` field in the results.';
+  override readonly sourceUrlSupport = 'package';
+  override readonly sourceUrlNote =
+    'The source url is determined by using the `packageName` and `registryUrl`.';
```
Comment (HonkingGoose) [inline lib/modules/datasource/github-tags/index.ts:29]:
```suggestion
    'The source URL is determined by using the `packageName` and `registryUrl`.';
```

File: lib/modules/datasource/gitlab-releases/index.ts [hunk 14]
```diff
@@ -11,6 +11,13 @@ export class GitlabReleasesDatasource extends Datasource {
 
   static readonly registryStrategy = 'first';
 
+  override readonly releaseTimestampSupport = true;
+  override readonly releaseTimestampNote =
+    'The release timestamp is determined from the `released_at` field in the results.';
+  override readonly sourceUrlSupport = 'package';
+  override readonly sourceUrlNote =
+    'The source url is determined by using the `packageName` and `registryUrl`.';
```
Comment (HonkingGoose) [inline lib/modules/datasource/gitlab-releases/index.ts:19]:
```suggestion
    'The source URL is determined by using the `packageName` and `registryUrl`.';
```

File: lib/modules/datasource/gitlab-tags/index.ts [hunk 15]
```diff
@@ -12,6 +12,13 @@ export class GitlabTagsDatasource extends Datasource {
 
   protected override http: GitlabHttp;
 
+  override readonly releaseTimestampSupport = true;
+  override readonly releaseTimestampNote =
+    'To get release timestamp we use the `created_at` field from the response.';
+  override readonly sourceUrlSupport = 'package';
+  override readonly sourceUrlNote =
+    'The source url is determined by using the `packageName` and `registryUrl`.';
```
Comment (HonkingGoose) [inline lib/modules/datasource/gitlab-tags/index.ts:20]:
```suggestion
    'The source URL is determined by using the `packageName` and `registryUrl`.';
```

File: lib/modules/datasource/go/index.ts [hunk 16]
```diff
@@ -31,6 +31,12 @@ export class GoDatasource extends Datasource {
 
   override readonly customRegistrySupport = false;
 
+  override readonly releaseTimestampSupport = true;
+  override readonly releaseTimestampNote = `In case release timestamp isn't already returned from respective datasource used to fetch the releases, it is determined from the \`Time\` field in the results.`;
```
Comment (HonkingGoose) [inline lib/modules/datasource/go/index.ts:35]: This sentence is hard to read. How about something like this:

> If the release timestamp is _not_ returned from the respective datasoure used to fetch the releases, then Renovate uses the `Time` field in the results instead.

File: lib/modules/datasource/go/index.ts [hunk 17]
```diff
@@ -31,6 +31,12 @@ export class GoDatasource extends Datasource {
...
+  override readonly sourceUrlSupport = 'package';
+  override readonly sourceUrlNote =
+    'The source url is determined from the `packageName` and `registryUrl`.';
```
Comment (HonkingGoose) [inline lib/modules/datasource/go/index.ts:38]:
```suggestion
    'The source URL is determined from the `packageName` and `registryUrl`.';
```

File: lib/modules/datasource/golang-version/index.ts [hunk 18]
```diff
@@ -32,6 +32,13 @@ export class GolangVersionDatasource extends Datasource {
 
   override readonly defaultVersioning = semverVersioningId;
 
+  override readonly releaseTimestampSupport = true;
+  override readonly releaseTimestampNote =
+    'The release timestamp is determined from the `Date` field in the results.';
+  override readonly sourceUrlSupport = 'package';
+  override readonly sourceUrlNote =
+    'We use the url https://github.com/golang/go.';
```
Comment (HonkingGoose) [inline lib/modules/datasource/golang-version/index.ts:40]:
```suggestion
    'We use the URL: https://github.com/golang/go.';
```

File: lib/modules/datasource/gradle-version/index.ts [hunk 19]
```diff
@@ -20,6 +20,13 @@ export class GradleVersionDatasource extends Datasource {
 
   override readonly registryStrategy = 'merge';
 
+  override readonly releaseTimestampSupport = true;
+  override readonly releaseTimestampNote =
+    'The release timestamp is determined from the `buildTime` field in the results.';
+  override readonly sourceUrlSupport = 'package';
+  override readonly sourceUrlNote =
+    'We use the url https://github.com/gradle/gradle.';
```
Comment (HonkingGoose) [inline lib/modules/datasource/gradle-version/index.ts:28]:
```suggestion
    'We use the URL https://github.com/gradle/gradle.';
```

File: lib/modules/datasource/helm/index.ts [hunk 20]
```diff
@@ -25,6 +25,13 @@ export class HelmDatasource extends Datasource {
 
   override readonly defaultVersioning = helmVersioning.id;
 
+  override readonly releaseTimestampSupport = true;
+  override readonly releaseTimestampNote =
+    'The release timstamp is determined from the `created` field in the results.';
+  override readonly sourceUrlSupport = 'package';
+  override readonly sourceUrlNote =
+    'The source url is determined from the `home` field or the `sources` field in the results.';
```
Comment (HonkingGoose) [inline lib/modules/datasource/helm/index.ts:33]:
```suggestion
    'The source URL is determined from the `home` field or the `sources` field in the results.';
```

File: lib/modules/datasource/hermit/index.ts [hunk 21]
```diff
@@ -30,6 +30,10 @@ export class HermitDatasource extends Datasource {
     'https://github.com/cashapp/hermit-packages',
   ];
 
+  override readonly sourceUrlSupport = 'release';
+  override readonly sourceUrlNote =
+    'The source url is determined from the `Repository` field in the results.';
```
Comment (HonkingGoose) [inline lib/modules/datasource/hermit/index.ts:35]:
```suggestion
    'The source URL is determined from the `Repository` field in the results.';
```

File: lib/modules/datasource/hex/index.ts [hunk 22]
```diff
@@ -18,6 +18,13 @@ export class HexDatasource extends Datasource {
 
   override readonly defaultVersioning = hexVersioning.id;
 
+  override readonly releaseTimestampSupport = true;
+  override readonly releaseTimestampNote =
+    'The relase timestamp is determined the `inserted_at` field in the results.';
```
Comment (HonkingGoose) [inline lib/modules/datasource/hex/index.ts:23]:
```suggestion
    'The release timestamp is determined from the `inserted_at` field in the results.';
```

Typo in the word `release`. And a missing word `from`.

File: lib/modules/datasource/hex/index.ts [hunk 23]
```diff
@@ -18,6 +18,13 @@ export class HexDatasource extends Datasource {
...
+  override readonly sourceUrlSupport = 'package';
+  override readonly sourceUrlNote =
+    'The source url from the `Github` field present in the results.';
```
Comment (HonkingGoose) [inline lib/modules/datasource/hex/index.ts:26]:
```suggestion
    'The source URL is determined from the `Github` field present in the results.';
```

I think this is the correct pattern?

File: lib/modules/datasource/hexpm-bob/index.ts [hunk 24]
```diff
@@ -24,6 +24,13 @@ export class HexpmBobDatasource extends Datasource {
 
   override readonly defaultVersioning = semverId;
 
+  override readonly releaseTimestampSupport = true;
+  override readonly releaseTimestampNote =
+    'The release timestamp is determined from the `buildDate` field in the results.';
+  override readonly sourceUrlSupport = 'package';
+  override readonly sourceUrlNote =
+    'We use the urls, https://github.com/elixir-lang/elixir.git for `elixir` and https://github.com/erlang/otp.git for `erlang`.';
```
Comment (HonkingGoose) [inline lib/modules/datasource/hexpm-bob/index.ts:32]:
```suggestion
    'We use these URLs: https://github.com/elixir-lang/elixir.git for `elixir`, and https://github.com/erlang/otp.git for `erlang`.';
```

How about this?

File: lib/modules/datasource/jenkins-plugins/index.ts [hunk 25]
```diff
@@ -25,6 +25,13 @@ export class JenkinsPluginsDatasource extends Datasource {
   private static readonly packageInfoPath = 'current/update-center.actual.json';
   private static readonly packageVersionsPath = 'current/plugin-versions.json';
 
+  override readonly releaseTimestampSupport = true;
+  override readonly releaseTimestampNote =
+    'The releaseTimestamp is determined from the `releaseTimestamp` or `buildDate` field in the results.';
+  override readonly sourceUrlSupport = 'package';
+  override readonly sourceUrlNote =
+    'The source url is determined from the `scm` field in the results.';
```
Comment (HonkingGoose) [inline lib/modules/datasource/jenkins-plugins/index.ts:33]:
```suggestion
    'The source URL is determined from the `scm` field in the results.';
```

File: lib/modules/datasource/maven/index.ts [hunk 26]
```diff
@@ -69,6 +69,13 @@ export class MavenDatasource extends Datasource {
 
   override readonly registryStrategy: RegistryStrategy = 'merge';
 
+  override readonly releaseTimestampSupport = true;
+  override readonly releaseTimestampNote =
+    'The release timestamp is determined from the `Last-Modified` header or the `lastModified` field in the results.';
+  override readonly sourceUrlSupport = 'package';
+  override readonly sourceUrlNote =
+    'The source url is determined from the `scm` tags in the results.';
```
Comment (HonkingGoose) [inline lib/modules/datasource/maven/index.ts:77]:
```suggestion
    'The source URL is determined from the `scm` tags in the results.';
```

File: lib/modules/datasource/node-version/index.ts [hunk 27]
```diff
@@ -19,6 +19,13 @@ export class NodeVersionDatasource extends Datasource {
 
   override readonly caching = true;
 
+  override readonly releaseTimestampSupport = true;
+  override readonly releaseTimestampNote =
+    'The release timestamp is determined from the `date` field.';
+  override readonly sourceUrlSupport = 'package';
+  override readonly sourceUrlNote =
+    'We use the url https://github.com/nodejs/node';
```
Comment (HonkingGoose) [inline lib/modules/datasource/node-version/index.ts:27]:
```suggestion
    'We use the URL: https://github.com/nodejs/node';
```

File: lib/modules/datasource/npm/index.ts [hunk 28]
```diff
@@ -14,6 +14,13 @@ export class NpmDatasource extends Datasource {
 
   override readonly defaultVersioning = npmVersioning.id;
 
+  override readonly releaseTimestampSupport = true;
+  override readonly releaseTimestampNote =
+    'The release timestamp is determined from the `time` field in the results.';
+  override readonly sourceUrlSupport = 'release';
+  override readonly sourceUrlNote =
+    'The source url is determined from the `repository` field in the results.';
```
Comment (HonkingGoose) [inline lib/modules/datasource/npm/index.ts:22]:
```suggestion
    'The source URL is determined from the `repository` field in the results.';
```

File: lib/modules/datasource/nuget/index.ts [hunk 29]
```diff
@@ -18,6 +18,17 @@ export class NugetDatasource extends Datasource {
 
   override readonly registryStrategy = 'merge';
 
+  override readonly releaseTimestampSupport = true;
+  override readonly releaseTimestampNote = `
+  For the v2 api, the release timestamp is determined from the \`Publised\` tag and,
```
Comment (HonkingGoose) [inline lib/modules/datasource/nuget/index.ts:23]:
```suggestion
  For the v2 API, the release timestamp is determined from the \`Published\` tag and,
```

I think this is a typo? Also capitalize abbreviations like API.

File: lib/modules/datasource/nuget/index.ts [hunk 30]
```diff
@@ -18,6 +18,17 @@ export class NugetDatasource extends Datasource {
...
+  for the v3 api, the release timestamp is determined from the \`published\` field in the results.
```
Comment (HonkingGoose) [inline lib/modules/datasource/nuget/index.ts:24]:
```suggestion
  for the v3 API, the release timestamp is determined from the \`published\` field in the results.
```

File: lib/modules/datasource/nuget/index.ts [hunk 31]
```diff
@@ -18,6 +18,17 @@ export class NugetDatasource extends Datasource {
...
+   `;
+  override readonly sourceUrlSupport = 'package';
+  override readonly sourceUrlNote = `
+  For the v2 api, the source url is determined from the \`ProjectUrl\` tag and,
```
Comment (HonkingGoose) [inline lib/modules/datasource/nuget/index.ts:28]:
```suggestion
  For the v2 API, the source URL is determined from the \`ProjectUrl\` tag and,
```

File: lib/modules/datasource/nuget/index.ts [hunk 32]
```diff
@@ -18,6 +18,17 @@ export class NugetDatasource extends Datasource {
...
+  for the v3 api, the source url is determined from the \`metadata.repository@url\` field in the results.
```
Comment (HonkingGoose) [inline lib/modules/datasource/nuget/index.ts:29]:
```suggestion
  for the v3 API, the source URL is determined from the \`metadata.repository@url\` field in the results.
```

Or it `sourceUrl`?

**Review by HonkingGoose — COMMENTED** [review]
Some common things to fix:

- Capitalizating abbreviations
- Grammar

Please check if my suggestions are good/bad before applying them.

Please also check the files I haven't commented on, I ran out of time. :smile:

**Review by secustor — APPROVED** [review]

**Review by rarkins — COMMENTED** [review]
Should we add this to our docs in this PR or a separate one?

**Comment by RahulGautamSingh** [comment]
Separate would be better, as that will be a doc thing so the scope will change.

**Comment by rarkins** [comment]
@RahulGautamSingh please create a follow up Issue then for integrating this with the docs

**Comment by HonkingGoose** [comment]
> @RahulGautamSingh please create a follow up Issue then for integrating this with the docs

I created issue:

- #29186

for us to track this work.

**Pull request merged**